app/models/application.rb

```diff
@@ -18,6 +18,19 @@ class Application < ApplicationRecord
   belongs_to :cohort, optional: true
 
   has_many :ecf_sync_request_logs, as: :syncable, dependent: :destroy
+  has_many :previously_funded_applications, lambda {
+    includes(:course).eligible_for_funding.accepted.where("EXISTS(
+    WITH json_data(alt_courses) AS (VALUES ('#{ActiveRecord::Base.sanitize_sql("{\"1\":[1],\"8\":[8,7],\"11\":[11],\"3\":[3],\"4\":[4],\"5\":[5],\"6\":[6],\"7\":[7,8],\"9\":[9],\"2\":[2],\"10\":[10]}")}'::jsonb))
```

Comment:
Let's revisit this next week. I don't think we should be relying on JSON in queries whenever possible.

Reply:
Having had a think, my understanding of the query above is this:

1. each course is mapped to itself except `7` and `8`, which overlap because they are "Additional Support Offer for new headteachers" and "Early Headship Coaching Offer" respectively, which sound similar so you probably don't qualify for funding on one if you've done the other 
    ```
    {
      1: [1],
      2:[2],
      3:[3],
      4:[4],
      5:[5],
      6:[6],
      7:[7,8],
      8:[8,7],
      9:[9],
      10:[10],
      11:[11],
      }
    ```
2. if a user has done the course (or a similar one in the case of `7` and `8`) they can't be funded for it again, so this matches and the record is returned

The thing that springs to mind is we should be able to solve this cleanly with data.

![image](https://github.com/DFE-Digital/npq-registration/assets/128088/6668f34f-0474-42a4-a68f-453557e9e836)

In the schema we added `CourseGroup` - this might have been intended for something else but a model _like_ this, where we had a "Headship" group that "Additional Support Offer for new headteachers" and "Early Headship Coaching Offer" belonged to, we could write the query without a CTE, without hardcoded IDs, without JSON  and without a subquery.

Reply:
So this came out of trying to move away from our solution in ECF to write a query for this which might be confusing. The original query in ECF is the following:
```
participant_identity
      .npq_applications
      .where.not(id:)
      .where(npq_course: npq_course.rebranded_alternative_courses)
      .where(eligible_for_funding: true)
      .accepted
      .exists?
```
which is basically checking if we have accepted applications that have been previously funded for the same course for the same user.
@leandroalemao let's chat about how we can maybe have something in the db which maybe is a flag that's set at certain stages of acceptance (which can come later) which we can just read on the record itself to confirm if someone has been accepted or not, as @peteryates mentioned attempting to solve with data itself. Give me a shout so we can chat :)